tools/linter/adapters/test_has_main_linter.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+"""
+This lint verifies that every Python test file (file that matches test_*.py or
+*_test.py in the test folder) has a main block which raises an exception or
+calls run_tests to ensure that the test will be run in OSS CI.
+
+Takes ~2 minuters to run without the multiprocessing, probably overkill.
+"""
+import argparse
+import json
+import multiprocessing as mp
+from enum import Enum
+from typing import List, NamedTuple, Optional
+
+import libcst as cst
+import libcst.matchers as m
+
+LINTER_CODE = "TEST_HAS_MAIN"
+
+
+class HasMainVisiter(cst.CSTVisitor):
+    def __init__(self) -> None:
+        super().__init__()
+        self.found = False
+
+    def visit_Module(self, node: cst.Module) -> bool:
+        name = m.Name("__name__")
+        main = m.SimpleString('"__main__"') | m.SimpleString("'__main__'")
+        run_test_call = m.Call(
+            func=m.Name("run_tests") | m.Attribute(attr=m.Name("run_tests"))
+        )
+        raise_block = m.Raise()
+
+        # name == main or main == name
+        if_main1 = m.Comparison(
+            name,
+            [m.ComparisonTarget(m.Equal(), main)],
+        )
+        if_main2 = m.Comparison(
+            main,
+            [m.ComparisonTarget(m.Equal(), name)],
+        )
+        for child in node.children:
+            if m.matches(child, m.If(test=if_main1 | if_main2)):
+                if len(m.findall(node, run_test_call | raise_block)) > 0:
```

Comment:
This should also work, right?

```suggestion
                if m.findall(node, run_test_call | raise_block):
```